exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,596 @@
+use storage::View;
+use to_handle;
+use utils::{
+    jni_cache::runtime_adapter,
+    panic_on_exception, unwrap_jni,
+};
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+    deployed_artifacts: HashSet<JavaArtifactId>,
+    started_services: HashMap<InstanceId, Instance>,
+    started_services_by_name: HashMap<String, InstanceId>,
+}
+
+/// Service identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone)]
+struct Instance {
+    id: InstanceId,
+    name: String,
+}
+
+/// Artifact identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct JavaArtifactId {
+    /// `groupId` maven property
+    pub group: String,
+    /// `artifactId` maven property
+    pub artifact: String,
+    /// `version` maven property in format {major.}{minor}.{patch}
+    pub version: Version,
+}
+
+struct AfterCommitContext<'a> {
+    dispatcher: &'a DispatcherSender,
+    snapshot: &'a dyn Snapshot,
+    service_keypair: &'a (PublicKey, SecretKey),
+    tx_sender: &'a ApiSender,
+}
+
+/// List of possible Java runtime errors.
+#[derive(Debug, Copy, Clone)]
+#[repr(u8)]
+pub enum Error {
+    /// Unable to parse artifact identifier or specified artifact has non-empty spec.
+    IncorrectArtifactId = 0,
+    /// Checked java exception is occurred
+    JavaException = 2,
+    /// Unspecified error
+    UnspecifiedError = 3,
+    /// Not supported operation
+    NotSupportedOperation = 4,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceStateHashes")]
+struct ServiceStateHashes {
+    instance_id: u32,
+    state_hashes: Vec<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Clone, ProtobufConvert, PartialEq)]
+#[exonum(pb = "proto::ServiceRuntimeStateHashes")]
+struct ServiceRuntimeStateHashes {
+    runtime_state_hashes: Vec<Vec<u8>>,
+    service_state_hashes: Vec<ServiceStateHashes>,
+}
+
+impl JavaRuntimeProxy {
+    /// Runtime Identifier
+    pub const RUNTIME_ID: RuntimeIdentifier = RuntimeIdentifier::Java;
+
+    /// Creates new `JavaRuntimeProxy` for given `ServiceRuntimeAdapter` object
+    pub fn new(executor: Executor, adapter: GlobalRef) -> Self {
+        JavaRuntimeProxy {
+            exec: executor,
+            runtime_adapter: adapter,
+            deployed_artifacts: HashSet::new(),
+            started_services: HashMap::new(),
+            started_services_by_name: HashMap::new(),
+        }
+    }
+
+    fn parse_artifact(&self, artifact: &ArtifactId) -> Result<JavaArtifactId, ExecutionError> {
+        if artifact.runtime_id != Self::RUNTIME_ID as u32 {
+            return Err(Error::IncorrectArtifactId.into());
+        }
+        artifact
+            .name
+            .parse()
+            .map_err(|_| Error::IncorrectArtifactId.into())
+    }
+
+    fn add_started_service(&mut self, instance: Instance) {
```

Comment:
Runtime must handle attempt to start already started service

Reply:
Not the runtime proxy.

Reply:
Okay, but then the list of started services must be stored in Java, mustn't it?

Reply:
How would you implement the service runtime without storing the services in Java?